src/Simplex/Messaging/Client.hs

```diff
@@ -389,25 +396,45 @@ getProtocolClient g transportSession@(_, srv, _) cfg@ProtocolClientConfig {qSize
+          lastRecv <- newTVarIO sessionTs
+          raceAny_ ([send c' th, process c', receive c' th lastRecv] <> [ping c' lastRecv | smpPingInterval > 0])
             `finally` disconnected c'
 
     send :: Transport c => ProtocolClient v err msg -> THandle v c 'TClient -> IO ()
     send ProtocolClient {client_ = PClient {sndQ}} h = forever $ atomically (readTBQueue sndQ) >>= \(active, s) -> whenM (readTVarIO active) (void $ tPutLog h s)
 
-    receive :: Transport c => ProtocolClient v err msg -> THandle v c 'TClient -> IO ()
-    receive ProtocolClient {client_ = PClient {rcvQ}} h = forever $ tGet h >>= atomically . writeTBQueue rcvQ
-
-    ping :: ProtocolClient v err msg -> IO ()
-    ping c@ProtocolClient {client_ = PClient {pingErrorCount}} = do
-      threadDelay' smpPingInterval
-      runExceptT (sendProtocolCommand c Nothing "" $ protocolPing @v @err @msg) >>= \case
-        Left PCEResponseTimeout -> do
-          cnt <- atomically $ stateTVar pingErrorCount $ \cnt -> (cnt + 1, cnt + 1)
-          when (maxCnt == 0 || cnt < maxCnt) $ ping c
-        _ -> ping c -- sendProtocolCommand resets pingErrorCount
+    receive :: Transport c => ProtocolClient v err msg -> THandle v c 'TClient -> TVar UTCTime -> IO ()
+    receive ProtocolClient {client_ = PClient {rcvQ}} h lastRecv = forever $ do
+      tGet h >>= atomically . writeTBQueue rcvQ
+      getCurrentTime >>= atomically . writeTVar lastRecv
+
+    ping :: ProtocolClient v err msg -> TVar UTCTime -> IO ()
+    ping c@ProtocolClient {client_ = PClient {sendPings, pingErrorCount, transportHost}} lastRecv = do
+      atomically $ unlessM (readTVar sendPings) retry
+      loop smpPingInterval
       where
+        loop :: Int64 -> IO ()
+        loop delay = do
+          threadDelay' delay
+          diff <- diffUTCTime <$> getCurrentTime <*> readTVarIO lastRecv
+          let idle = diffToMicroseconds diff
+              remaining = smpPingInterval - idle
+          if remaining > 250_000 -- delay pings only for significant time
+            then do
+              logDebug $ thost <> " : last response was " <> tshow diff <> " ago, delaying ping for " <> tshow (secondsToNominalDiffTime $ fromIntegral remaining / 1000000)
+              loop remaining
+            else do
+              logDebug $ thost <> " : last response was " <> tshow diff <> " ago, sending ping"
+              -- sendProtocolCommand/getResponse updates pingErrorCount
+              runExceptT (sendProtocolCommand c Nothing "" $ protocolPing @v @err @msg) >>= \case
+                Left PCEResponseTimeout -> do
+                  cnt <- readTVarIO pingErrorCount
+                  -- drop client when maxCnt of commands (pings or otherwise) have timed out in sequence, but only after some time has passed after last received response
+                  when (maxCnt == 0 || cnt < maxCnt || diff < recoverWindow) $ loop smpPingInterval
```

Comment:
the problem with this implementation is that if PING is never sent and the client keeps trying to send messages and they all timeout, the client will never be dropped.

Reply:
most likely the check to exit is needed in the first branch too